---
title: "10x_complete.myeloid"
author: "Sandra Tietscher"
date: "10/31/2019"
output: html_document
---

## Load libraries and data

```{r}
if (!require(ggplot2)) install.packages('ggplot2')
library(ggplot2)

if (!require(dplyr)) install.packages('dplyr')
library(dplyr)

if (!require(Seurat)) install.packages('Seurat')
library(Seurat)

if (!require(tidyverse)) install.packages('tidyverse')
library(tidyverse)

if (!require(data.table)) install.packages('data.table')
library(data.table)

if (!require(sctransform)) install.packages('sctransform')
library(sctransform)

if(!require(clustree)) install.packages("clustree")
library(clustree)

library(magrittr)
library(RColorBrewer)
#library(plyr) -> wilcoxon testing line does not work when this is loaded after dplyr
library(corrplot)
library(pheatmap)
library(ComplexHeatmap)
library(scales)
library(viridis)
library(circlize)

in.path = "~/Code/10x/10x_Run1_Run2_Pilot_comb/output/subset_objects/"
out.path = "~/Code/10x/10x_Run1_Run2_Pilot_comb/output/myeloid/"

heatmap.color <- brewer.pal(n=10, name= "RdBu")
heatmap.color.rev <- rev(heatmap.color)

complete.myeloid <- readRDS(file= paste0(out.path, "myeloid_17dim.rds"))
```

## Load object

```{r}
complete.myeloid <- readRDS(file = paste(in.path, "myeloid_inclDC.rds", sep = ''))

# store Keratin and MGP percentage in object meta data
complete.myeloid <- PercentageFeatureSet(complete.myeloid, pattern = "^KRT", col.name = "percent.krt")
complete.myeloid <- PercentageFeatureSet(complete.myeloid, pattern = "MGP", col.name = "percent.MGP")

table(complete.myeloid@meta.data$orig.ident)
table(complete.myeloid$TIG)
```

## Apply sctransform normalization

This replaces FindVariableFeatures, Normalize and ScaleData. 
```{r}

# run sctransform
#complete.myeloid <- SCTransform(complete.myeloid , vars.to.regress = c("percent.mt"), verbose = TRUE)
complete.myeloid <- SCTransform(complete.myeloid , vars.to.regress = c("percent.mt", "percent.krt", "percent.MGP"), verbose = TRUE)
```

## Re-Run Dimensional Reduction Analysis
```{r}

#### Run PCA ###
complete.myeloid <- RunPCA(object = complete.myeloid, verbose = FALSE)

print(x = complete.myeloid[['pca']], dims = 1:5, nfeatures = 5, projected = FALSE)
VizDimLoadings(object = complete.myeloid, dims = 1:2)
PCAPlot(object = complete.myeloid)

#Choose significant dimensions
DimHeatmap(object = complete.myeloid, dims = 15, cells = 500, balanced = TRUE)
DimHeatmap(object = complete.myeloid, dims = 10:20, cells = 500, balanced = TRUE)
ElbowPlot(object = complete.myeloid, ndims =30)


### run UMAP ###

complete.myeloid <- RunUMAP(object = complete.myeloid, dims = 1:17)

# note that you can set `label = TRUE` or use the LabelClusters function to help label individual clusters
DimPlot(object = complete.myeloid, reduction = 'umap', label = TRUE)
DimPlot(object = complete.myeloid, reduction = 'umap', group.by = "orig.ident")
DimPlot(object = complete.myeloid, reduction = 'umap', group.by = "run")
DimPlot(object = complete.myeloid, reduction = 'umap', group.by = "TIG")

DimPlot(object = complete.myeloid, reduction = 'umap', group.by = "orig.ident", split.by="orig.ident", ncol = 7)

```

## Graph-based clustering

```{r}
complete.myeloid <- FindNeighbors(object = complete.myeloid, dims = 1:17)
complete.myeloid <- FindClusters(object = complete.myeloid, resolution = 0.8)

#Query number of clusters and cells per cluster
cells.per.cluster <- table(... = complete.myeloid@meta.data$SCT_snn_res.0.8)
cells.per.cluster
write.csv(cells.per.cluster, file = paste0(out.path, "17dim_res0.8CellsPerCluster.csv"), row.names = FALSE)

complete.myeloid$seurat_clusters <- complete.myeloid$SCT_snn_res.0.8
Idents(complete.myeloid) <- complete.myeloid$SCT_snn_res.0.8

# QC plots
VlnPlot(object = complete.myeloid, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size = 0, sort = FALSE)
VlnPlot(object = complete.myeloid, features = c("MKI67"), pt.size = 0, sort = FALSE)
VlnPlot(object = complete.myeloid, features = c("percent.krt", "percent.MGP"), pt.size = 0, sort = FALSE)
```

## Clustree analysis

```{r}

p = clustree(complete.myeloid, prefix = "SCT_snn_res.", exprs = "scale.data")

```

### Remove low-quality clusters
Remove tiny cluster 16 (high mitochondrial percentage).

```{r}
Idents(complete.myeloid) <- complete.myeloid$SCT_snn_res.0.8
complete.myeloid<- subset(x=complete.myeloid, idents = 0:15)

#Save object
saveRDS(complete.myeloid, file = paste(out.path, "myeloid_17dim.rds", sep = ''))
```

## Cluster proportions per sample & per TIG

```{r} 

#Which clusters is each sample made of?
cluster.sample.prop <- prop.table(x=table(Idents(object=complete.myeloid), complete.myeloid$orig.ident), margin=2)
write.csv(cluster.sample.prop, file = paste0(out.path, "cluster_samples_proportions.csv"), row.names = TRUE)
cluster.sample.prop <- as.data.frame(cluster.sample.prop)
colnames(cluster.sample.prop) <- c('cluster', 'sample', 'prop')

p = ggplot(cluster.sample.prop, aes(sample, y = prop, fill=cluster)) +
  geom_bar(stat="identity")+
  theme(axis.title.x=element_blank(), axis.title.y = element_blank())+
  coord_flip()+
  theme(panel.background = element_blank())+
  ggtitle('Sample composition (T/NK clusters)')
ggsave(file=file.path(out.path,paste0('sample_composition.png')),width=10, height=6,dpi=300)


### per sample ###

cluster.sample = table(Idents(object=complete.myeloid), complete.myeloid$orig.ident)
#write.csv(cluster.sample, file = paste0(out.path, "cluster_samples.csv"), row.names = TRUE)

sample.cluster.prop <- prop.table(cluster.sample, margin = 1)
write.csv(sample.cluster.prop, file = paste0(out.path, "sample_cluster_proportions.csv"), row.names = TRUE)

#prepare data table for stacked barplot
sample.cluster.prop <- as.data.frame(sample.cluster.prop)
colnames(sample.cluster.prop) <- c('cluster', 'sample', 'prop')

# Which samples is each cluster made of?
p = ggplot(sample.cluster.prop, aes(cluster, y = prop, fill=sample)) +
  geom_bar(stat="identity")+
  theme(axis.title.x=element_blank(), axis.title.y = element_blank())+
  coord_flip()+
  theme(panel.background = element_blank())+
  ggtitle('T/NK cell cluster composition')
ggsave(file=file.path(out.path,paste0('cluster_composition_sample.png')),width=8, height=4,dpi=300)

# How many cells of each sample are in each cluster?
cluster.sample <- as.data.frame(cluster.sample)
colnames(cluster.sample) <- c("cluster", "sample", "cell.nr")
p = ggplot(cluster.sample, aes(cluster, y = cell.nr, fill=sample)) +
  geom_bar(stat="identity")+
  theme(axis.title.x=element_blank(), axis.title.y = element_blank())+
  #coord_flip()+
  theme(panel.background = element_blank())+
  ggtitle('T/NK cell cluster composition')
ggsave(file=file.path(out.path,paste0('cluster_totalcells_sample.png')),width=8, height=4,dpi=300)


#What percentage of each sample is in each cluster?
cluster.sample.prop <- as.data.frame(cluster.sample.prop)
cluster.sample.prop$TIG <- ifelse(cluster.sample.prop$sample == "TBB129" | cluster.sample.prop$sample  == "TBB171" | cluster.sample.prop$sample  == "TBB111" | cluster.sample.prop$sample  == "TBB165" | cluster.sample.prop$sample  == "TBB338" | cluster.sample.prop$sample  == "TBB184" | cluster.sample.prop$sample  == "TBB011", "TIG2", "TIG3")
colnames(cluster.sample.prop) <- c("cluster", "sample", "proportion", "TIG")
cluster.sample.prop$sample <- factor(cluster.sample.prop$sample, levels = c("TBB129", "TBB171" , "TBB111" ,"TBB165", "TBB338", "TBB011", "TBB184", "TBB075", "TBB102", "TBB214", "TBB226", "TBB330", "TBB212", "TBB035"))

#Barplots
p = ggplot(tdat, aes(x= sample, y = proportion, fill=TIG)) +
  geom_bar(stat="identity", position = "dodge")+
  theme(axis.title.x=element_blank(), axis.title.y = element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank())+
  facet_wrap(~cluster, scales = "fixed", ncol = 6)+
  #coord_flip()+
  theme(panel.background = element_blank())
  #ggtitle('Percentage of sample x that falls in cluster y')
ggsave(file=file.path(out.path,paste0('barplot_cluster_sampleprop.png')),width=8, height=4,dpi=300)

### per TIG ###

cluster.TIG = table(complete.myeloid$SCT_snn_res.0.8, complete.myeloid$TIG)
write.csv(cluster.TIG, file = paste0(out.path, "cluster_TIG.csv"), row.names = TRUE)
cluster.TIG.prop <- prop.table(x=table(Idents(object=complete.myeloid), complete.myeloid$TIG), margin=2)
write.csv(cluster.TIG.prop, file = paste0(out.path, "cluster_TIG_proportion.csv"), row.names = TRUE)

TIG.cluster.prop <- prop.table(cluster.TIG, margin = 1)
write.csv(TIG.cluster.prop, file = paste0(out.path, "TIG_cluster_proportions.csv"), row.names = TRUE)

#What percentage of each TIG is in which cluster? (Grouped barplot)
cluster.TIG.prop<- as.data.frame(cluster.TIG.prop)
colnames(cluster.TIG.prop) <- c('cluster', 'TIG', 'prop')
p = ggplot(cluster.TIG.prop, aes(cluster, y = prop, fill=TIG)) +
  geom_bar(stat="identity", position = "dodge")+
  theme(axis.title.x=element_blank(), axis.title.y = element_blank())+
  #coord_flip()+
  theme(panel.background = element_blank())+
  ggtitle('Proportion of each TIG belonging to a specific cluster')
ggsave(file=file.path(out.path,paste0('TIG_cluster_prop.png')),width=8, height=4,dpi=300)


### Calculate Wilcoxon Test
tdat <- as.data.table(cluster.sample.prop)
tdat$cluster <- as.factor(tdat$cluster)
tdat$TIG <- as.factor(tdat$TIG)
tdat$sample <- as.factor(tdat$sample)
pvalues <- tdat %>% 
  group_by(cluster) %>% 
  summarise(p=wilcox.test(proportion~TIG)$p.value)
tdat <- merge(tdat, pvalues, by.x = "cluster", by.y ="cluster", all.x = TRUE)
tdat$p.wt <- paste0('p=',round(tdat$p, digits=3))

#Boxplots
p = ggplot(tdat, aes(x= TIG, y = proportion, fill=TIG))+
  geom_boxplot()+
  geom_point()+
  facet_wrap(~cluster+p.wt, scales="free", ncol=8, strip.position="bottom")+
  theme(axis.title.x=element_text("cluster"), axis.text.x=element_blank(), axis.ticks.x = element_blank())+
  #theme(panel.background = element_blank())+
  xlab("cluster")
ggsave(file=file.path(out.path,paste0('boxplot_cluster_wilcoxon.png')),width=12, height=5,dpi=300)



#How many cells of each TIG are in each cluster?
cluster.TIG <- as.data.frame(cluster.TIG)
colnames(cluster.TIG) <- c("cluster", "TIG", "cell.nr")

p = ggplot(cluster.TIG, aes(cluster, y = cell.nr, fill=TIG)) +
  geom_bar(stat="identity")+
  theme(axis.title.x=element_blank(), axis.title.y = element_blank())+
  #coord_flip()+
  theme(panel.background = element_blank())+
  ggtitle('T/NK cell cluster composition')

ggsave(file=file.path(out.path,paste0('cluster_TIG.png')),width=8, height=4,dpi=300)

ggsave(paste0(out.path,"FeaturePlots/cytokine_receptor.png"), plot = p, width = 400, height = 250, units = "mm")


### Proportions of each cluster corrected for total proportion of myeloid cells 
celltype.sample.prop <- read.csv("~/Code/10x/10x_Run1_Run2_Pilot_comb/output/celltype_samples_proportions_inclImmune.csv", row.names=1)
metacluster.sample.prop <- prop.table(x=table(complete.myeloid$metacluster, complete.myeloid$orig.ident), margin=2)
myeloid.prop <- as.numeric(celltype.sample.prop[2,])
metacluster.total.prop <- sweep(metacluster.sample.prop, MARGIN=2, myeloid.prop, `*`)
immune.prop <- as.numeric(celltype.sample.prop["immune",])
metacluster.immune.prop <- sweep(metacluster.total.prop, MARGIN=2, immune.prop, `/`)

#What percentage of each sample is in each cluster?
metacluster.total.prop <- as.data.frame(metacluster.total.prop)
colnames(metacluster.total.prop) <- c("metacluster", "sample", "proportion")
metacluster.total.prop$TIG <- ifelse(metacluster.total.prop$sample == "TBB129" | metacluster.total.prop$sample  == "TBB171" | metacluster.total.prop$sample  == "TBB111" | metacluster.total.prop$sample  == "TBB165" | metacluster.total.prop$sample  == "TBB338" | metacluster.total.prop$sample  == "TBB184" | metacluster.total.prop$sample  == "TBB011", "TIG2", "TIG3")
metacluster.total.prop$sample <- factor(metacluster.total.prop$sample, levels = c("TBB129", "TBB171" , "TBB111" ,"TBB165", "TBB338", "TBB011", "TBB184", "TBB075", "TBB102", "TBB214", "TBB226", "TBB330", "TBB212", "TBB035"))

### Calculate Wilcoxon Test
tdat <- as.data.table(metacluster.total.prop)
tdat$metacluster <- as.factor(tdat$metacluster)
tdat$TIG <- as.factor(tdat$TIG)
tdat$sample <- as.factor(tdat$sample)
pvalues <- tdat %>% 
  group_by(metacluster) %>% 
  summarise(p=wilcox.test(proportion~TIG)$p.value)
tdat <- merge(tdat, pvalues, by.x = "metacluster", by.y ="metacluster", all.x = TRUE)
tdat$p.wt <- paste0('p=',round(tdat$p, digits=3))

#Boxplots
p = ggplot(tdat, aes(x= TIG, y = proportion, fill=TIG))+
  geom_boxplot()+
  geom_point()+
  facet_wrap(~metacluster+p.wt, scales="free", ncol=5, strip.position="bottom")+
  theme(axis.title.x=element_text("metacluster"), axis.text.x=element_blank(), axis.ticks.x = element_blank())+
  #theme(panel.background = element_blank())+
  xlab("metacluster")+
  ylab("Proportion of all immune cells")
ggsave(file=file.path(out.path,paste0('boxplot_cluster_wilcoxon.png')),width=12, height=5,dpi=300)

#Total percentage corrected for immune cell percentage
metacluster.immune.prop <- as.data.frame(metacluster.immune.prop)
colnames(metacluster.immune.prop) <- c("metacluster", "sample", "proportion")
metacluster.immune.prop$TIG <- ifelse(metacluster.immune.prop$sample == "TBB129" | metacluster.immune.prop$sample  == "TBB171" | metacluster.immune.prop$sample  == "TBB111" | metacluster.immune.prop$sample  == "TBB165" | metacluster.immune.prop$sample  == "TBB338" | metacluster.immune.prop$sample  == "TBB184" | metacluster.immune.prop$sample  == "TBB011", "TIG2", "TIG3")
metacluster.immune.prop$sample <- factor(metacluster.immune.prop$sample, levels = c("TBB129", "TBB171" , "TBB111" ,"TBB165", "TBB338", "TBB011", "TBB184", "TBB075", "TBB102", "TBB214", "TBB226", "TBB330", "TBB212", "TBB035"))

### Calculate Wilcoxon Test
tdat <- as.data.table(metacluster.immune.prop)
tdat$metacluster <- as.factor(tdat$metacluster)
tdat$TIG <- as.factor(tdat$TIG)
tdat$sample <- as.factor(tdat$sample)
pvalues <- tdat %>% 
  group_by(metacluster) %>% 
  summarise(p=wilcox.test(proportion~TIG)$p.value)
tdat <- merge(tdat, pvalues, by.x = "metacluster", by.y ="metacluster", all.x = TRUE)
tdat$p.wt <- paste0('p=',round(tdat$p, digits=3))
```

## Finding differentially expressed features


```{r}

# find markers for every cluster compared to all remaining cells, report only the positive ones. Save list as csv.
Idents(complete.myeloid) <- complete.myeloid$SCT_snn_res.0.8
complete.myeloid.markers.clean <- FindAllMarkers(object = complete.myeloid, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
write.csv(complete.myeloid.markers.clean, file = paste0(out.path, "dim15_res1_cluster_AllMarkerGenes_clean.csv"), row.names = FALSE)
marker.genes <- complete.myeloid.markers.clean %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
write.csv(marker.genes, file = paste0(out.path, "dim15_res1_cluster_MarkerGenesTop10_clean.csv"), row.names = TRUE)

# using MAST
Idents(complete.myeloid) <- complete.myeloid$SCT_snn_res.0.8
markers.clean.MAST <- FindAllMarkers(object = complete.myeloid, only.pos = TRUE, min.pct = 0.05, logfc.threshold = 0.25, test.use="MAST")
write.csv(markers.clean.MAST, file = paste0(out.path, "DE_metacluster_AllMarkerGenes_MAST.csv"), row.names = FALSE)
marker.genes <- markers.clean.MAST %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
write.csv(marker.genes, file = paste0(out.path, "DE_cluster_MarkerGenesTop10_clean_MAST.csv"), row.names = TRUE)


# compare two groups
M1M2.TAM.MAST <- FindMarkers(object = complete.myeloid, ident.1="M1M2_TAM", ident.2 = "TAM", min.pct = 0.05, logfc.threshold = 0.25, test.use="MAST")
write.csv(M1M2.TAM.MAST, file = paste0(out.path, "DE_M1M2vsTAM_MAST.csv"), row.names = TRUE)


#specific cluster markers
#Idents(complete.myeloid) <- complete.myeloid$seurat_clusters
#naive.markers.MAST <- FindMarkers(object = complete.myeloid, ident.1=9, min.pct = 0.25, logfc.threshold = 0.25, test.use="MAST")
#write.csv(naive.markers.MAST, file = paste0(in.path, "DE_cluster9_MAST.csv"), row.names = TRUE)
```

## Feature lists ##
```{r}
features = c("FOXP3", "CCL18",  "IL2RA")
features2 = c("FLT1")
features_cytof = c('CD3E', 'CD8A', 'CD4', 'FOXP3', 'HAVCR2', 'PDCD1', 'CTLA4', 'ICOS', 'IL2RA', 'PTPRC', 'CD68', 'CD14', 'CD274', 'CCR7', 'HLA-DRA', 'MRC1', 'SIGLEC1', 'MSR1', 'CD163', 'FCGR2A', 'FCGR2B', 'FCGR2C','FCGR1A', 'ITGAM', 'ITGAX', 'FCGR3A', 'CD93', 'IL3RA', 'CD86', 'CD36', 'CD38', 'CCR2', 'SLAMF7')
features_cytof_TAM = c('CD68', 'CD14', 'CD274', 'CCR7', 'HLA-DRA', 'MRC1', 'SIGLEC1', 'MSR1', 'CD163', 'FCGR2A', 'FCGR2B', 'FCGR2C', 'FCGR1A', 'CD4', 'ITGAM', 'ITGAX', 'FCGR3A', 'CD93', 'IL3RA', 'CD86', 'CD36', 'CD38', 'CCR2', 'SLAMF7', "IRF8", "S100A9", "CD33")
features_cytof_T = c('CD3E', 'CD8A', 'FOXP3', 'HAVCR2', 'PDCD1', 'CTLA4','ICOS', 'IL2RA', 'PTPRC', 'CD4', 'CCR7', 'CD38')
chemokine = c('CCL20', 'CCL22', 'CXCL2', 'CXCL3', 'CXCL8', 'CCL8', 'CCL18', 'CCL2', 'CCL3', 'CCL4', 'CCL4L2', 'CXCL10', 'CXCL12', 'CCL13', 'CXCL1', 'CXCL13')
cytokine= c('CCL20', 'CCL22', 'CXCL2', 'CXCL3', 'CXCL8', 'CCL8', 'CCL18', 'CCL2', 'CCL3', 'CCL4', 'CCL4L2', 'CCL5', 'CXCL10', 'CXCL12', 'CCL13', 'CXCL1', 'CXCL13', "IL4", "IL10", "IL13", "IFNG", "TNF")
chemokine_01 = c('CCL20', 'CCL22', 'CXCL2', 'CXCL3', 'CXCL8')
chemokine_02 = c('CCL8', 'CCL18', 'CCL2', 'CCL3', 'CCL4')
chemokine_03 = c( 'CCL4L2', 'CXCL10', 'CXCL12', 'CCL13', 'CXCL1')
cytokine_receptor = c('CCR1','CCR10','CCR2','CCR7', 'CCR4', 'CCR5', 'CCR6', 'IL10RA', 'IL4R', 'CXCR2', 'CXCR3', 'CXCR4', 'CXCR5')
TF = c('IRF2', 'IRF5', 'IRF8', 'IRF9', 'IRF4', 'IRF7','STAT1', 'STAT2', 'STAT4', 'TCF12', 'TCF19', 'BCL6',
       'ZBTB31', 'ZBTB33', 'ZBTB47', 'CIITA')
bulk_GOI = c('CD163', 'CD209', 'DAB2', 'F13A1', 'FOLR2', 'MAF', 'MS4A4A', 'SLC38A6', 'LYVE1', 'PDK4', 'APOBEC3A', 'VCAN', 'CD1C', 'CD300E', 'CD1E', "S100A12")
bulk_GOI_topDE = c("FCN1", "EREG", "CD300E", "S100A8", "APOBEC3A", "AREG", "FCER1A", "CD52", "THBS1", "WARS", "VCAN", 'CD1C', 'CD1E', "S100A12", "FOLR2", "CCL8", "F13A1", "MRC1", "SLC40A1", "ME1", "VSIG4", "CCL18", "CCL13", 'CD163', 'CD209', 'DAB2', 'FOLR2', 'MAF', 'MS4A4A', 'SLC38A6', 'LYVE1', 'PDK4')

MMPs <- c("MMP1", "MMP2", "MMP3", "MMP7", "MMP8", "MMP9", "MMP10", "MMP11", "MMP12", "MMP13", "MMP14", "MMP15", "MMP16", "MMP17", "MMP19", "MMP21", "MMP23B", "MMP24", "MMP25", "MMP26", "MMP27", "MMP28")

type1_cytokines = c("IFNG", "IL12A", "IL12B", "TNF")
type2_cytokines = c("IL4", "IL6", "IL10", "IL13")
angiogenic_factors = c("VEGFA", "VEGFB", "LYVE1", "PLAU", "ADM", "TYMP","TEK", "FGF2")
#complement_activation = as.vector(read.csv(paste0(out.path, "HGNC_complement_system_activation_components_list.csv"))[,1])
complement_red = c("C1QA", "C1QB", "C1QC", "C2", "C3")


##DC subset markers from Villani et al (Science 2017)
# DCmarkers <- as.data.table(read.csv(paste0(out.path, "VillaniScience2017_DC_subset_markers_top15.csv")))
# DCmarkers <- DCmarkers[Gene.ID != "AK128525" & Gene.ID != "IGJ" & Gene.ID != "C1ORF54" & Gene.ID !="C12ORF75" & Gene.ID !=  "AK307192" & Gene.ID != "HLA-DQB"  & Gene.ID != "LOC645638" & Gene.ID != "HLA-DPB2"] #remove genes that are not present in my myeloid dataset
# DCmarkers.ind <- as.vector(DCmarkers$Gene.ID)

##M1M2
M1M2_ligands <- c("IL10", "HGF", "LPL", "CXCL12", "CCL13", "IGF1", "TGFB2", "SPP1", "APOE", "C3", "ICAM5", "CGA", "ANOS1", "LAMA1")
M1M2_receptors <- c("CALCR", "KCNQ3", "ITGB3", "NRP1", "SDC3", "EDNRB")

aDC <- c("CD274", "IDO1", "IL15", "FSCN1", "LAMP3", "CCR7", "CCL17", "CCL19", "CCL22", "CD80", "AIRE", "FOXD4")
aDC.red <- c("CD274", "IDO1", "CD80", "FSCN1", "LAMP3", "CCR7", "IL15",  "CCL17", "CCL19", "CCL22")
```

## Cluster averages ##

```{r}

Idents(complete.myeloid) <- complete.myeloid$SCT_snn_res.0.8
cluster.averages <- AverageExpression(object = complete.myeloid, return.seurat = TRUE)
cluster.averages

#Calculate averages for metaclusters
Idents(complete.myeloid) <- complete.myeloid$metacluster
metacluster.averages <- AverageExpression(object = complete.myeloid, return.seurat = TRUE)
metacluster.averages

#Calculate averages for samples
Idents(complete.myeloid) <- complete.myeloid$orig.ident
sample.averages <- AverageExpression(object = complete.myeloid, return.seurat = TRUE)
sample.averages

# Add celltype information
#cluster.averages[['cluster']] <- Idents(object=cluster.averages)
#cluster.averages <- RenameIdents(object = cluster.averages, new.cluster.ids)
#Save celltype information in metadata
#cluster.averages [["cell.type"]] <- Idents(object = cluster.averages )
#Reset active IDs to original clusters
#Idents(cluster.averages) <- cluster.averages[["cluster"]]


## Calculate expression averages separately for each sample
# Careful #: sample order gets mixed up, ordered by cluster size (not always the same order for each sample)
cluster.average.clean.TIG <- AverageExpression(object=complete.myeloid, return.seurat=TRUE, add.ident = "TIG")
cluster.average.clean.TIG$TIG

### Heatmap of chosen features (cluster averages)
#total
p = DoHeatmap(object = sample.averages, features = "CD1C", size = 5, draw.lines = FALSE, angle=45)
  #scale_fill_gradientn(colors = colorRampPalette(c("#9900CC", "#070707", "#FFFF33"))(256))
  #scale_fill_gradientn(colors = heatmap.color.rev)
  
ggsave(paste0(out.path,"heatmaps/MMP_metallothioneins.png"), plot = p, width = 400, height = 250, units = "mm")  

#single cell heatmap
myeloid.small <- subset(x = complete.myeloid, idents = c("cycling", "unknown"), invert=TRUE)
myeloid.small$metacluster <- droplevels(myeloid.small$metacluster)
p = DoHeatmap(subset(myeloid.small), features = aDC.red, size = 3, draw.lines = TRUE, lines.width = 200) + scale_fill_viridis(100)

#by TIG
p = DoHeatmap(object = cluster.average.clean.TIG, features = features_T_extended, size = 3, draw.lines = FALSE, angle=0)+
  scale_fill_gradientn(colors = heatmap.color.rev)

```

### Complex heatmaps ####

```{r}
cluster.ids <- read.csv(file = paste0(out.path, "myeloid_metacluster_assignment_v2.csv"))

cluster.sample = table(complete.myeloid$SCT_snn_res.0.8, complete.myeloid$orig.ident)
cluster.sample <- cluster.sample[1:16,]
my_colors <- hue_pal()(14)
heatmap_colors = colorRampPalette(c("#9900CC", "#070707", "#FFFF33"))(256)

#Return as tables
Idents(complete.myeloid) <- complete.myeloid$metacluster
cluster.averages.table <- AverageExpression(object = myeloid.small, return.seurat = FALSE)
RNA.average <- as.matrix(cluster.averages.table[[1]])
RNA.average.scaled <- as.matrix(cluster.averages.table[[2]])
#Normalize between 0 and 1
RNA.average.norm <- apply(RNA.average, 1, function(x)(x-min(x))/(max(x)-min(x)))
RNA.average.znorm <- apply(RNA.average, 1, function(x)(x - mean(x)) / sd(x))

#Reorder
#RNA.average.norm <- RNA.average.norm[c(6,3,1, 5,4,2),]
RNA.average.znorm <- RNA.average.znorm[-4,]

#just the heatmap
p = Heatmap(subset(RNA.average.znorm, select=aDC), show_row_names = TRUE, row_names_side = "right", row_dend_side = "left", heatmap_legend_param = list(title = "Z-scored\nmean counts"), col=viridis(100))
p = Heatmap(subset(RNA.average.znorm, select=aDC.red), show_row_names = TRUE, row_names_side = "left", column_names_side = "top", column_names_rot = 70, cluster_rows = FALSE, cluster_columns = FALSE, heatmap_legend_param = list(title = "Z-scored mean counts", direction = "horizontal"), col=viridis(100))

#construct stacked barplot for row annotation and the corresponding legend
row_ha <- rowAnnotation(cells = row_anno_barplot(cluster.sample, axis_param=list(side="bottom"), gp=gpar(fill=my_colors), width = unit(4, "cm")), annotation_name_side = "bottom")
lgd <- Legend(at = colnames(cluster.sample), title = "patient", type = "points", legend_gp = gpar(col=my_colors))

p = Heatmap(subset(RNA.average.znorm, select=features_cytof_TAM), show_row_names = TRUE, row_names_side = "right", row_dend_side = "left", heatmap_legend_param = list(title = "Z-scored intensity", direction = "horizontal"), right_annotation = row_ha, col=magma(100))
#add legend
p1 = draw(p, annotation_legend_list = list(lgd), heatmap_legend_side = "bottom")

#construct row annotation for metaclusters
cluster.ids <- read.csv(file = paste0(out.path, "myeloid_metacluster_assignment_v2.csv"))
metacl_colors <- hue_pal()(9)
names(metacl_colors) <- levels(cluster.ids$metacluster)
metacl_ha <- rowAnnotation(metacluster = cluster.ids$metacluster, col = list(metacluster= metacl_colors))
lgd2 <- Legend(at = levels(cluster.ids$metacluster), title = NULL, type = "point", legend_gp = gpar(col=metacl_colors))

p = Heatmap(subset(RNA.average.norm, select=features_cytof_TAM), show_row_names = TRUE, row_names_side = "right", row_dend_side = "left", heatmap_legend_param = list(title = "Normalized intensity", direction = "horizontal"), right_annotation = metacl_ha, col=viridis(100))


## heatmap with marker annotation (e.g. DC subset marker) at the side

#construct column annotation for metaclusters
DCpop <- DCmarkers$DC_subset
pop_colors <- hue_pal()(6)
names(pop_colors) <- levels(DCmarkers$DC_subset)
DCsubset_ha <- HeatmapAnnotation(DC_subset = DCmarkers$DC_subset, col = list(DC_subset= pop_colors))

p2 = Heatmap(subset(RNA.average.znorm, select=DCmarkers.ind), show_row_names = TRUE, row_names_side = "right", row_dend_side = "left", cluster_columns =  FALSE, heatmap_legend_param = list(title = "Normalized intensity", direction = "horizontal"), top_annotation = DCsubset_ha, col = heatmap_colors)
```

## Big DE heatmap

```{r}
#marker.genes <- read.csv(file = paste0(out.path, "DE_cluster_MarkerGenesTop10_clean_MAST.csv"))

all.marker.genes <- read.csv(file = paste0(out.path, "DE_cluster_AllMarkerGenes_clean_MAST.csv"))
marker.genes <- all.marker.genes %>% group_by(cluster) %>% top_n(n = 5, wt = avg_logFC)

features <- marker.genes$gene
features.cluster <- marker.genes[,6:7]
colnames(features.cluster) <- c('cluster_expressed', 'gene')
# remove duplicate genes
features.cluster <- features.cluster[!duplicated(features.cluster$gene), ]


p = DoHeatmap(object = cluster.averages, features = features, size = 0, draw.lines = FALSE, angle=45)+
  scale_fill_gradientn(colors = magma(100))

average_matrix <- GetAssayData(cluster.averages[["SCT"]], slot="scale.data")
average_df<- as.data.frame(average_matrix)
average_df <- rownames_to_column(average_df, var = "gene")
average_df <- gather(average_df, key = "cluster", value = "scaled_average", 2:17)
levels(average_df$cluster) <- factor(0:15)

#use only top10 DE features of each cluster
average_df <- average_df[average_df$gene %in% features,]
big.DE <- merge(average_df, features.cluster, by = 'gene')

#Prepare order of variables
big.DE$cluster <- as.numeric(big.DE$cluster)
big.DE$cluster <- as.factor(big.DE$cluster)
big.DE$cluster <- factor(big.DE$cluster, levels=c(17:0))
big.DE$gene <- as.factor(big.DE$gene)
big.DE$gene <- factor(big.DE$gene, levels = unique(big.DE$gene[order(big.DE$cluster_expressed)]))


#ggplot heatmap
p = ggplot(big.DE, aes(x=gene, y=cluster))+
  geom_tile(aes(fill=scaled_average))+
  theme(axis.title.y = element_blank(), axis.text.x = element_text(angle = 90, hjust = 1, size=10), axis.text.y = element_text(size=12))+
  scale_fill_gradientn(colors = magma(100))
ggsave(paste0(out.path, "Heatmap_Top10genes_cluster.png"))
```

## Assigning subtype and role identity to clusters

```{r}

#Read in table with cluster ids
cluster.ids <- read.csv(file = paste0(out.path, "myeloid_metacluster_assignment.csv"))
Idents(complete.myeloid) <- complete.myeloid$SCT_snn_res.0.8

subtype.ids <- cluster.ids$metacluster
subtype.ids <- as.character(subtype.ids)
names(x = subtype.ids) <- levels(x = complete.myeloid)
complete.myeloid <- RenameIdents(object = complete.myeloid, subtype.ids)
#Save celltype information in metadata
complete.myeloid[["metacluster"]] <- Idents(object = complete.myeloid)

complete.myeloid$metacluster <- factor(complete.myeloid$metacluster,levels(complete.myeloid$metacluster)[c(5,3,2,8,7,9,1,6,4)])

#Reset active IDs to original clusters
#Idents(complete.myeloid) <- complete.myeloid[["SCT_snn_res.0.8"]]

# plot_celltype 
DimPlot(object = complete.myeloid, reduction = 'umap', label = FALSE, pt.size = 0.5, group.by = "metacluster")
ggsave(paste0(out.path, "umap_metacluster.png"), width=10, height=7)

```

### Compare metaclusters
by sample and by TIG

```{r}
#Which metaclusters is each sample made of?
metacluster.sample.prop <- prop.table(x=table(complete.myeloid$metacluster, complete.myeloid$orig.ident), margin=2)
write.csv(metacluster.sample.prop, file = paste0(out.path, "metacluster_samples_proportions.csv"), row.names = TRUE)
metacluster.sample.prop <- as.data.frame(metacluster.sample.prop)
colnames(metacluster.sample.prop) <- c('metacluster', 'sample', 'prop')

p = ggplot(metacluster.sample.prop, aes(sample, y = prop, fill=metacluster)) +
  geom_bar(stat="identity")+
  theme(axis.title.x=element_blank(), axis.title.y = element_blank())+
  coord_flip()+
  theme(panel.background = element_blank())+
  ggtitle('Sample composition (Myeloid metaclusters)')
ggsave(file=file.path(out.path,paste0('sample_composition_metaclusters_wholeTAM.png')),width=10, height=6,dpi=300)



### What percentage of each sample/TIG is in each metacluster? ###

metacluster.sample.prop.df <- as.data.frame(metacluster.sample.prop)
metacluster.sample.prop.df$TIG <- ifelse(metacluster.sample.prop.df$sample == "TBB129" | metacluster.sample.prop.df$sample  == "TBB171" | metacluster.sample.prop.df$sample  == "TBB111" | metacluster.sample.prop.df$sample  == "TBB165" | metacluster.sample.prop.df$sample  == "TBB338" | metacluster.sample.prop.df$sample  == "TBB184" | metacluster.sample.prop.df$sample  == "TBB011", "TIG2", "TIG3")
#arrange samples by TIG
metacluster.sample.prop.df$sample <- factor(metacluster.sample.prop.df$sample, levels=unique(metacluster.sample.prop.df$sample[order(metacluster.sample.prop.df$TIG)]), ordered=TRUE)
#order metaclusters
metacluster.sample.prop.df$metacluster <- factor(metacluster.sample.prop.df$metacluster, levels = c("pDC", "cDC", "cDC_CD1C+",  "cDC_CLEC9A+", "DC_MDSC?", "mono_early", "TAM", "cytof_TAM", "cycling", "unknown"))

#Barplots
p = ggplot(metacluster.sample.prop.df, aes(x= sample, y = prop, fill=TIG)) +
  geom_bar(stat="identity", position = "dodge")+
  theme(axis.title.x=element_blank(), axis.title.y = element_blank(), axis.text.x=element_blank())+
  facet_wrap(~metacluster, scales = "free")+
  theme(panel.background = element_blank())
ggsave(file=file.path(out.path,paste0('metacluster_per_sample.png')),width=8, height=4,dpi=300)


### Calculate Wilcoxon Test
tdat <- metacluster.sample.prop.df
pvalues <- tdat %>% 
  group_by(metacluster) %>% 
  summarise(p=wilcox.test(prop~TIG)$p.value)
tdat <- merge(tdat, pvalues, by.x = 1, by.y =1, all.x = TRUE)
tdat$p.wt <- paste0('p=',round(tdat$p, digits=3))

#Boxplots (with stats)
p = ggplot(tdat, aes(x= TIG, y = prop, fill=TIG))+
  geom_boxplot()+
  geom_point()+
  facet_wrap(~metacluster+p.wt, scales="free", ncol=5, strip.position="top")+
  theme(axis.ticks.x = element_blank(), 
        axis.text.x=element_blank(), 
        axis.text.y = element_text(size=12),
        axis.title.x = element_blank(), 
        #panel.border = element_rect(colour = "black", fill=NA, size=1),
        axis.line.x = element_line(colour = "black", size = 0.25),
        axis.line.y = element_line(colour = "black", size = 0.25),
        panel.background = element_blank())+
  ylab("proportion")
ggsave(file=file.path(out.path,paste0('metacluster_per_TIG_wilcoxon.png')),width=8, height=4,dpi=300)
```
### Differential expression between metaclusters
```{r}

# using MAST
Idents(complete.myeloid) <- complete.myeloid$metacluster
markers.MAST <- FindMarkers(object = complete.myeloid, ident.1="cytof_TAM", ident.2 = "TAM", test.use="MAST")
write.csv(markers.MAST, file = paste0(out.path, "DE_cytofTAM_vs_TAM_MAST.csv"), row.names = TRUE)

Idents(complete.myeloid) <- complete.myeloid$metacluster
markers.MAST <- FindMarkers(object = complete.myeloid, ident.1="cytof_TAM", test.use="MAST")
write.csv(markers.MAST, file = paste0(out.path, "DE_cytofTAM_vs_all_MAST.csv"), row.names = TRUE)


```


## Prepare data for pseudobulk analysis 
For cluster-independent TIG comparison

```{r}

##### without computing sample averages
raw.counts <- complete.myeloid@assays$RNA@counts
raw.counts <- as.data.frame(raw.counts)
raw.counts$TBB011 <- rowSums(raw.counts[,grep("TBB011", names(raw.counts))])
raw.counts$TBB035 <- rowSums(raw.counts[,grep("TBB035", names(raw.counts))])
raw.counts$TBB075 <- rowSums(raw.counts[,grep("TBB075", names(raw.counts))])
raw.counts$TBB102 <- rowSums(raw.counts[,grep("TBB102", names(raw.counts))])
raw.counts$TBB111 <- rowSums(raw.counts[,grep("TBB111", names(raw.counts))])
raw.counts$TBB129 <- rowSums(raw.counts[,grep("TBB129", names(raw.counts))])
raw.counts$TBB165 <- rowSums(raw.counts[,grep("TBB165", names(raw.counts))])
raw.counts$TBB171 <- rowSums(raw.counts[,grep("TBB171", names(raw.counts))])
raw.counts$TBB184 <- rowSums(raw.counts[,grep("TBB184", names(raw.counts))])
raw.counts$TBB212 <- rowSums(raw.counts[,grep("TBB212", names(raw.counts))])
raw.counts$TBB214 <- rowSums(raw.counts[,grep("TBB214", names(raw.counts))])
raw.counts$TBB226 <- rowSums(raw.counts[,grep("TBB226", names(raw.counts))])
raw.counts$TBB330 <- rowSums(raw.counts[,grep("TBB330", names(raw.counts))])
raw.counts$TBB338 <- rowSums(raw.counts[,grep("TBB338", names(raw.counts))])

raw.sums <- raw.counts[,27357:27370]
write.csv(raw.sums, file = paste0(out.path, "sample_sum_counts.csv"), row.names = TRUE)


### for DCs only (excl. pDCs)
Idents(complete.myeloid) <- complete.myeloid$metacluster
DC.only <- subset(complete.myeloid, idents = c("cDC", "cDC1", "cDC2", "DC_MDSC"))
#run code from above
raw.sums <- raw.counts[,6429:6442]
write.csv(raw.sums, file = paste0(out.path, "sample_sum_counts_DC.csv"), row.names = TRUE)

### for TAMs only 
Idents(complete.myeloid) <- complete.myeloid$metacluster
TAM.only <- subset(complete.myeloid, idents = "TAM")
#run code from above
raw.sums <- raw.counts[,13170:13183]
write.csv(raw.sums, file = paste0(out.path, "sample_sum_counts_TAM.csv"), row.names = TRUE)
```

## Look more into cluster 11 (DC_MDSC)

```{r}
#subset the cluster
Idents(complete.myeloid) <- complete.myeloid$metacluster
cl11 <- subset(x = complete.myeloid, idents = "DC_MDSC")
table(cl11$orig.ident)
#exclude TBB035 and TBB075, because they have <10 cells in this cluster and could lead to bias. TBB102 has no cells and is already excluded
Idents(cl11) <- cl11$orig.ident
cl11 <- subset(x= cl11, invert=TRUE, idents = c("TBB035", "TBB075"))

#prepare data for pseudobulk analysis
##### without computing sample averages
raw.counts <- cl11@assays$RNA@counts
raw.counts <- as.data.frame(raw.counts)
raw.counts$TBB011 <- rowSums(raw.counts[,grep("TBB011", names(raw.counts))])
raw.counts$TBB111 <- rowSums(raw.counts[,grep("TBB111", names(raw.counts))])
raw.counts$TBB129 <- rowSums(raw.counts[,grep("TBB129", names(raw.counts))])
raw.counts$TBB165 <- rowSums(raw.counts[,grep("TBB165", names(raw.counts))])
raw.counts$TBB171 <- rowSums(raw.counts[,grep("TBB171", names(raw.counts))])
raw.counts$TBB184 <- rowSums(raw.counts[,grep("TBB184", names(raw.counts))])
raw.counts$TBB212 <- rowSums(raw.counts[,grep("TBB212", names(raw.counts))])
raw.counts$TBB214 <- rowSums(raw.counts[,grep("TBB214", names(raw.counts))])
raw.counts$TBB226 <- rowSums(raw.counts[,grep("TBB226", names(raw.counts))])
raw.counts$TBB330 <- rowSums(raw.counts[,grep("TBB330", names(raw.counts))])
raw.counts$TBB338 <- rowSums(raw.counts[,grep("TBB338", names(raw.counts))])

raw.sums <- raw.counts[,1126:1136]
write.csv(raw.sums, file = paste0(out.path, "cluster11_sample_sum_counts.csv"), row.names = TRUE)

## feature plots
GOI <- c("CD274", "LAMP3", "BIRC3", "CCL19", "CCL22", "IDO1", "CD40", "FSCN1", "CD80", "CXCL9", "IL15", "CCL17", "CCR7")
FeaturePlot(complete.myeloid, features = GOI)

#####################################
## new PCA, UMAP and subclustering ##
#####################################

cl11 <- RunPCA(object = cl11, verbose = FALSE)

print(x = cl11[['pca']], dims = 1:5, nfeatures = 5, projected = FALSE)
VizDimLoadings(object = cl11, dims = 1:2)
PCAPlot(object = cl11)

#Choose significant dimensions
DimHeatmap(object = cl11, dims = 1:10, cells = 500, balanced = TRUE)
ElbowPlot(object = cl11, ndims =10)

### run UMAP ###
cl11 <- RunUMAP(object = cl11, dims = 1:10)
DimPlot(object = cl11, reduction = 'pca', label = TRUE)
DimPlot(object = cl11, reduction = 'pca', group.by = "orig.ident")
DimPlot(object = cl11, reduction = 'pca', group.by = "TIG")

#Cluster
cl11 <- FindNeighbors(object = cl11, dims = 1:10)
cl11 <- FindClusters(object = cl11, resolution = 0.2)

#Differentially expressed genes between the two clusters
cl11.sub <- FindAllMarkers(object = cl11, only.pos = TRUE, min.pct = 0.1, logfc.threshold = 0.25)
```

## Look more into proliferating cells

```{r}
#subset the cluster
Idents(complete.myeloid) <- complete.myeloid$metacluster
prolif <- subset(x = complete.myeloid, idents = "Proliferating")
table(prolif$orig.ident)

#####################################
## new PCA, UMAP and subclustering ##
#####################################

prolif <- RunPCA(object = prolif, verbose = FALSE)

print(x = prolif[['pca']], dims = 1:5, nfeatures = 5, projected = FALSE)
VizDimLoadings(object = prolif, dims = 1:2)
PCAPlot(object = prolif)

#Choose significant dimensions
DimHeatmap(object = prolif, dims = 1:10, cells = 500, balanced = TRUE)
ElbowPlot(object = prolif, ndims =10)

### run UMAP ###
prolif <- RunUMAP(object = prolif, dims = 1:8)
DimPlot(object = prolif, reduction = 'pca', label = TRUE)
DimPlot(object = prolif, reduction = 'pca', group.by = "orig.ident")
DimPlot(object = prolif, reduction = 'pca', group.by = "TIG")

#Cluster
prolif <- FindNeighbors(object = prolif, dims = 1:8)
prolif <- FindClusters(object = prolif, resolution = 0.2)

DimPlot(prolif)
DimPlot(prolif, group.by  = "TIG")
FeaturePlot(prolif, c("CD68", "CCR7", "CCR2", "MRC1"))

#Differentially expressed genes between the 4 clusters
prolif.sub <- FindAllMarkers(object = prolif, only.pos = TRUE, min.pct = 0.1, logfc.threshold = 0.25)
```


